### DISCUSSION TITLE
Additional context keys for tree items

### DISCUSSION THREAD

**Discussion by vaclavHala** [microsoft/vscode-discussions#836] [Extension Development QnA] [answered]
Hi, I'm trying to get this use case to work
* I have tree view which shows as top level items projects open in the workspace which are of some special type (what this type is and how to recognize which projects are of it does not matter here).
* Each project has as children some items. The items can be of different well known types. Set of the types is static and known upfront in the extension.
* The items have some commands associated with them rendered as inline buttons in the item. Each item type has different commands. Which commands to show for given type of item I use the `viewItem` context key
* Based on current state of the project some of the commands for given type of item in given project may be disabled - this is the problem. 

The view should look something like this (notice `cmd2` is disabled for `type1` in `project1` and  `cmd4` is disabled for `type2` in `project2` )
```
> project1
    type1        cmd1
    type1        cmd1
    type2        cmd2 cmd3 cmd4
> project2
    type1        cmd1 cmd2
    type2        cmd2 cmd3


// full type1: cmd1 cmd2
// full type2: cmd2 cmd3 cmd4 
```

I want to write `when` clause for each of the inline commands like this `view == my-tree-view && viewItem == item-type-2 && project in enabled.item-type-2` where
* `enabled.item-type-2` is list containing URIs of projects for which items of `item-type-2` should be enabled - this list changes dynamically at runtime
* `project` should be URI of the project particular item belongs to - I can't find a way to set this (I'd either need option to have multiple keys per item or single key with structured value)

What I tried and doesn't work
* using `resourceUri` of the item to create some synthetic URI like `myUriType://${project}/${task}` so I could use e.g. `resourceDirname` in the `when` clause - this does not work because even when evaluating context for tree item all the `resource*` keys seem to have value of resource in the active editor (according to `Developer: Inspect Context Keys`) 
* using some composite string containing both item type and parent project in `item.contextValue` - this does not work because the `in` operator in `when` seems to only work for looking up exact match of string inside array, not for checking if one string is contained in other string

I'll be glad for any tips, thanks

**Comment by r3m0t** [comment]
Either: use an array of strings with `resourceUri  in myarray` and update the array often

Or: use regex match, eg `://type[134]` or you can encode multiple flags, like `typeSNC`/`typeUNC` and use regex `type.N.` where say, S = saved, U = unsaved, N = neat, C = colourful - the regex can match attributes and use `.` for attributes it doesn't care about. You can put the flags in the query or fragment part of the uri if you are already using the path part for something important.

**Reply by vaclavHala** [reply]
So are you saying I can somehow access `resourceUri` of individual tree items in `when` expressions? I tried this and couldn't get it to work, how can I do it?

I tried having a look into vscode sources and if I understand this part correctly https://github.com/microsoft/vscode/blob/main/src/vs/workbench/browser/parts/views/treeView.ts#L1549 only the `contextValue` of each tree item is added as new context key. So to access resourceUri of the item I'd need that code to include e.g.
```js
		const contextKeyService = this.contextKeyService.createOverlay([
			['view', this.id],
			['viewItem', element.contextValue],
         		['viewItemResourceUti', element.resourceUri],
		]);
```

**Reply by r3m0t** [reply, marked as answer]
Ah right, yes I mean `viewItem in myarray` and update myarray to have the contextValue of each tree item that the action should show on. You could use `treeItem.contextValue = resourceUri.toString()` if that's convenient.

**Reply by vaclavHala** [reply]
Oh I see what you mean, I can flip the way I use the context key and just give "ID" to each project/type combination and then manage the arrays which determine if cmd should be active for each "ID" like this
```
CTX showCmd1 = [project1/type1, project2/type2]
CTX showCmd2 = [project1/type2, project2/type1, project2/type2]
CTX showCmd3 = [project1/type2, project2/type2]
CTX showCmd4 = [project1/type2]

WHEN cmd1 = view == my-tree-view && viewItem not in showCmd1
WHEN cmd2 = view == my-tree-view && viewItem not in showCmd2
...
```
Now I can have whatever logic to decide how the individual arrays are filled

I didn't think about doing it this way and it works great for me, thanks
